src/optim/sgd.rs

```diff
@@ -244,6 +276,54 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_sgd_weight_decay_no_momentum() {
+        let mut sgd = Sgd::new(SgdConfig {
+            lr: 1e-2,
+            momentum: None,
+            weight_decay: Some(1e-3),
+        });
+
+        let mut t: Tensor1D<5> = Tensor1D::ones();
+        let rate = Tensor1D::new([0.1, 1.0, 2.0, 10.0, 100.0]);
+        let expected = [
+            [0.9988, 0.997, 0.995, 0.979, 0.799],
+            [0.99760115, 0.994003, 0.990005, 0.958021, 0.59820104],
+            [0.9964036, 0.991009, 0.98501503, 0.937063, 0.39760286],
+            [0.9952072, 0.988018, 0.98003, 0.9161259, 0.19720526],
+            [0.994012, 0.98502994, 0.97505, 0.8952098, -0.0029919297],
+        ];
+        for e in expected.iter() {
+            let gradients = backward((t.trace() * &rate).mean());
+            sgd.update(&mut t, gradients).expect("");
+            assert_eq!(t.data(), e);
```

Comment:
For non-exact results you can use `crate::tests::assert_close` instead of `assert_eq!`. Have been slowly moving tests to use this instead, but all the optimizers still use assert_eq

Reply:
Added #259 for this